README.md
# CloudLoggingUtil

[![Test](https://github.com/koshilife/cloud-logging-util/workflows/Test/badge.svg)](https://github.com/koshilife/cloud-logging-util/actions?query=workflow%3ATest)
[![codecov](https://codecov.io/gh/koshilife/cloud-logging-util/branch/master/graph/badge.svg)](https://codecov.io/gh/koshilife/cloud-logging-util)
[![Gem Version](https://badge.fury.io/rb/cloud-logging-util.svg)](http://badge.fury.io/rb/cloud-logging-util)
[![license](https://img.shields.io/github/license/koshilife/cloud-logging-util)](https://github.com/koshilife/cloud-logging-util/blob/master/LICENSE.txt)

CloudLoggingUtil is a small library to output structured logs in Cloud Logging.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'cloud-logging-util'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install cloud-logging-util

## Usage

Insert CloudLoggingUtil::AccessLogging on the head of rack middlewares.

### Rails

```rb
# config/application.rb
class Application < Rails::Application
  # setting for Access Logging ($stdout)
  config.middleware.insert_before(0, CloudLoggingUtil::AccessLogging)

  # setting for Application Logger ($stderr)
  config.logger = ActiveSupport::Logger.new($stderr)
  config.logger.formatter = CloudLoggingUtil::Formatter.new
  config.colorize_logging = false
end
```

Middleware check.

```sh
bundle exec rake middleware
```

setup.

```rb
class YourController < ActionController::Base
  prepend_before_action do
    # set trace_id to request_id
    CloudLoggingUtil.setup_trace_id(request.request_id)
  end
end
```

More in-depth method documentation can be found at [RubyDoc.info](https://www.rubydoc.info/gems/cloud-logging-util/).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/koshilife/cloud-logging-util. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/koshilife/cloud-logging-util/blob/master/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the Cloud::Logging::Logger project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/koshilife/cloud-logging-util/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [0.0.1.alpha1] - 2021-10-14

- Initial release


test/test_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require "simplecov"
SimpleCov.start
if ENV["CI"] == "true"
  require "codecov"
  SimpleCov.formatter = SimpleCov::Formatter::Codecov
end

require "minitest/autorun"
require "rack/test"
require "cloud-logging-util"

module CloudLoggingUtil
  class BaseTest < Minitest::Test
  end

  class RackTest < BaseTest
    include Rack::Test::Methods

    attr_reader :app

    def setup
      @mock_io = StringIO.new
      @app = AccessLogging.new(generate_app, io: @mock_io)
    end

    def generate_app
      raise NotImplementedError
    end
  end
end


test/cloud-logging-util/fomatter_test.rb
# frozen_string_literal: true

require "test_helper"

require "securerandom"
require "stringio"

module CloudLoggingUtil
  class FormatterTest < BaseTest
    def setup
      @dummy_io = StringIO.new
      @logger = Logger.new(@dummy_io)
      @logger.progname = "foobar_progname"
      @logger.formatter = Formatter.new
      @trace_id = SecureRandom.uuid
      CloudLoggingUtil.setup_trace_id(@trace_id)
    end

    def teardown
      CloudLoggingUtil.setup_trace_id(nil)
    end

    def test_it_that_logs_to_io
      @logger.info("foobar1")
      @logger.warn("foobar2") { { foo: 1, bar: 2 } }
      logs = @dummy_io.string.split("\n").map { |line| JSON.parse(line) }

      log = logs[0]
      assert(log.delete("timestamp"))
      expected = {
        "severity" => Severity::INFO,
        "progname" => "foobar_progname",
        "message" => "foobar1",
        "logging.googleapis.com/trace" => @trace_id
      }
      assert_equal(expected, log)

      log = logs[1]
      assert(log.delete("timestamp"))
      expected = {
        "severity" => Severity::WARNING,
        "progname" => "foobar2",
        "foo" => 1,
        "bar" => 2,
        "logging.googleapis.com/trace" => @trace_id
      }
      assert_equal(expected, log)
    end
  end
end


test/cloud-logging-util/severity_test.rb
# frozen_string_literal: true

require "test_helper"

module CloudLoggingUtil
  class SeverityTest < BaseTest
    def setup
      @subject = Severity
      @levels = ::Logger::SEV_LABEL
    end

    #
    # test for Severity.cast
    #

    def test_that_it_converts_debug_level
      assert_equal(@subject::DEBUG, @subject.cast(@levels[::Logger::DEBUG]))
    end

    def test_that_it_converts_info_level
      assert_equal(@subject::INFO, @subject.cast(@levels[::Logger::INFO]))
    end

    def test_that_it_converts_warn_level
      assert_equal(@subject::WARNING, @subject.cast(@levels[::Logger::WARN]))
    end

    def test_that_it_converts_error_level
      assert_equal(@subject::ERROR, @subject.cast(@levels[::Logger::ERROR]))
    end

    def test_that_it_converts_fatal_level
      assert_equal(@subject::CRITICAL, @subject.cast(@levels[::Logger::FATAL]))
    end

    def test_that_it_converts_any_level
      assert_equal(@subject::DEFAULT, @subject.cast(@levels[::Logger::UNKNOWN]))
      assert_equal(@subject::DEFAULT, @subject.cast("ANY"))
    end

    #
    # test for Severity.cast_from_http_status
    #

    def test_that_it_converts_empty_http_status
      assert_equal(@subject::ERROR, @subject.cast_from_http_status(nil))
    end

    def test_that_it_converts_http_status399
      assert_equal(@subject::INFO, @subject.cast_from_http_status(399))
    end

    def test_that_it_converts_http_status400
      assert_equal(@subject::WARNING, @subject.cast_from_http_status(400))
    end

    def test_that_it_converts_http_status500
      assert_equal(@subject::ERROR, @subject.cast_from_http_status(500))
    end
  end
end


lib/cloud-logging-util/log_entry.rb
# frozen_string_literal: true

module CloudLoggingUtil
  # Cloud Logging LogEntry
  # refs: https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry
  class LogEntry
    def initialize(severity, time, progname, message, trace_id: nil, additional_hash: nil)
      @severity = severity
      @time = time
      @progname = progname
      @message = message
      @trace_id = trace_id
      @additional_hash = additional_hash
    end

    def to_hash
      entry = {}
      entry.merge!(severity_h)
      entry.merge!(time_h)
      entry.merge!(progname_h)
      entry.merge!(message_h)
      entry.merge!(trace_h)
      @additional_hash.respond_to?(:merge) ? entry.merge(@additional_hash) : entry
    end

    private

    def severity_h
      return {} if @severity.nil?

      { severity: @severity }
    end

    def time_h
      return {} if @time.nil?
      return {} unless @time.respond_to?(:utc) && @time.respond_to?(:strftime)

      { timestamp: @time.utc.strftime("%FT%T.%NZ") }
    end

    def progname_h
      return {} if @progname.nil?

      { progname: @progname }
    end

    def message_h
      return {} if @message.nil?
      return @message.to_hash if @message.respond_to?(:to_hash)
      return { message: @message } if @message.is_a?(String)

      { message: @message.inspect }
    end

    def trace_h
      return {} if @trace_id.nil?

      {
        "logging.googleapis.com/trace": @trace_id
      }
    end
  end
end


lib/cloud-logging-util/access_logging.rb
# frozen_string_literal: true

require "time"
require "json"

require_relative "log_entry"
require_relative "severity"

module CloudLoggingUtil
  # Access Logging Rack Middleware for Cloud Logging
  class AccessLogging
    DEFAULT_PARAMS_PROC = proc do |env, status, headers, _body, began_at|
      now = Time.now
      latency = now.instance_eval { to_i + (usec / 1_000_000.0) } - began_at
      severity = Severity.cast_from_http_status(status)
      http_h = http_request_hash(env, status, headers, latency)
      LogEntry.new(severity, now, nil, nil, trace_id: trace_id(env), additional_hash: http_h).to_hash
    end

    def initialize(app, **kwargs)
      @app = app
      @io = kwargs[:io] || $stdout
      @params_proc = kwargs[:params_proc] || DEFAULT_PARAMS_PROC
    end

    def call(env)
      began_at = Time.now.instance_eval { to_i + (usec / 1_000_000.0) }
      status, headers, body = @app.call(env)
    ensure
      params = @params_proc.call(env, status, headers, body, began_at)
      @io.write("#{params.to_json}\n") if @io.respond_to?(:write)
    end

    def self.trace_id(env)
      env["cloud_logging_util.trace_id"] || env["action_dispatch.request_id"]
    end

    # refs: https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry#httprequest
    def self.http_request_hash(env, status, headers, latency)
      {
        httpRequest: {
          requestMethod: request_method(env),
          requestUrl: request_url(env),
          requestSize: request_size(env),
          status: status,
          responseSize: response_size(env, headers),
          userAgent: user_agent(env),
          remoteIp: remote_ip(env),
          serverIp: server_ip(env),
          referer: referer(env),
          latency: "%0.6fs" % latency,
          cacheLookup: cache_lookup(env),
          cacheHit: cache_hit(env),
          cacheValidatedWithOriginServer: cache_validated_with_origin_server(env),
          cacheFillBytes: cache_fill_bytes(env),
          protocol: protocol(env)
        }
      }
    end

    def self.request_method(env)
      if env["cloud_logging_util.request_method"].nil?
        env["REQUEST_METHOD"]
      else
        env["cloud_logging_util.request_method"]
      end
    end

    def self.request_url(env)
      if env["cloud_logging_util.request_url"].nil?
        url = "#{env["rack.url_scheme"]}://#{env["HTTP_HOST"]}#{env["PATH_INFO"]}"
        url += "?#{env["QUERY_STRING"]}" if env["QUERY_STRING"]&.length.to_i.positive?
        url
      else
        env["cloud_logging_util.request_url"].to_s
      end
    end

    def self.request_size(env)
      if env["cloud_logging_util.request_size"].nil?
        env["CONTENT_LENGTH"] ? env["CONTENT_LENGTH"].to_s : nil
      else
        env["cloud_logging_util.request_size"].to_s
      end
    end

    def self.response_size(env, headers)
      if env["cloud_logging_util.response_size"].nil?
        headers && headers["Content-Length"] ? headers["Content-Length"].to_s : nil
      else
        env["cloud_logging_util.response_size"].to_s
      end
    end

    def self.user_agent(env)
      env["cloud_logging_util.user_agent"] || env["HTTP_USER_AGENT"]
    end

    def self.remote_ip(env)
      env["cloud_logging_util.remote_ip"] || env["X-Forwarded-For"]
    end

    def self.server_ip(env)
      env["cloud_logging_util.server_ip"] || env["REMOTE_ADDR"]
    end

    def self.referer(env)
      env["cloud_logging_util.referer"] || env["HTTP_REFERER"]
    end

    def self.cache_lookup(env)
      if env["cloud_logging_util.cache_lookup"].nil?
        false
      else
        env["cloud_logging_util.cache_lookup"]
      end
    end

    def self.cache_hit(env)
      if env["cloud_logging_util.cache_hit"].nil?
        false
      else
        env["cloud_logging_util.cache_hit"]
      end
    end

    def self.cache_validated_with_origin_server(env)
      if env["cloud_logging_util.cache_validated_with_origin_server"].nil?
        false
      else
        env["cloud_logging_util.cache_validated_with_origin_server"]
      end
    end

    def self.cache_fill_bytes(env)
      if env["cloud_logging_util.cache_fill_bytes"].nil?
        nil
      else
        env["cloud_logging_util.cache_fill_bytes"].to_s
      end
    end

    def self.protocol(env)
      if env["cloud_logging_util.protocol"].nil?
        env["HTTP_VERSION"]
      else
        env["cloud_logging_util.protocol"]
      end
    end
  end
end


lib/cloud-logging-util/severity.rb
# frozen_string_literal: true

module CloudLoggingUtil
  # https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry#logseverity
  module Severity
    DEFAULT = "DEFAULT"
    DEBUG = "DEBUG"
    INFO = "INFO"
    NOTICE = "NOTICE"
    WARNING = "WARNING"
    ERROR = "ERROR"
    CRITICAL = "CRITICAL"
    ALERT = "ALERT"
    EMERGENCY = "EMERGENCY"

    def self.cast(severity)
      case severity.to_s.downcase
      when "debug"
        DEBUG
      when "info"
        INFO
      when "warn"
        WARNING
      when "error"
        ERROR
      when "fatal"
        CRITICAL
      else
        DEFAULT
      end
    end

    def self.cast_from_http_status(status)
      return ERROR if status.nil?

      status_int = status.to_s[0..2].to_i
      if status_int < 400
        INFO
      elsif status_int < 500
        WARNING
      else
        ERROR
      end
    end
  end
end
